### PR TITLE
Gz complex optimizations

### PR DESCRIPTION
**Pull request by gabuzi**

### CODE REVIEW

File: src/main/kotlin/org/vitrivr/cottontail/model/values/Complex32VectorValue.kt [hunk 1]
```diff
@@ -128,31 +127,34 @@ inline class Complex32VectorValue(val data: FloatArray) : ComplexVectorValue<Flo
...
         else -> FloatArray(this.data.size) {
             if (it % 2 == 0) {
-                this.data[it] + other[it].value.toFloat()
+                this.data[it] + other[it / 2].value.toFloat()
             } else {
                 this.data[it]
             }
         }
-    })
+    }) else throw DimensionMismatchException(this.logicalSize, other.logicalSize)
```
[the remark this answers is not in the document]
Comment (gabuzi) [reply src/main/kotlin/org/vitrivr/cottontail/model/values/Complex32VectorValue.kt:141]: 1. I ran some quick tests with the L1 distance on Complex64VV of different dimensions. The check is not noticeable.
2. I am in favor of creating a dedicated exception. I just used the one from Commons Math for a quick implementation.

File: src/main/kotlin/org/vitrivr/cottontail/math/knn/metrics/AbsoluteInnerProductSimilarity.kt [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    // todo: assess actual cost
+
+    /**
+     * Calculates the asolute value of the inner product between two [VectorValue]s.
+     *
+     * @param a First [VectorValue]
+     * @param b Second [VectorValue]
+     * @return Absolute value of inner product between a and b.
+     */
+    override fun invoke(a: VectorValue<*>, b: VectorValue<*>): DoubleValue = (a dot b).abs().asDouble()
```
[the remark this answers is not in the document]
Comment (gabuzi) [reply src/main/kotlin/org/vitrivr/cottontail/math/knn/metrics/AbsoluteInnerProductSimilarity.kt]: fixed with 4eab89ced6777709317856d04003a9f74b96028a

File: src/main/kotlin/org/vitrivr/cottontail/model/values/Complex32VectorValue.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (ppanopticon) [reply src/main/kotlin/org/vitrivr/cottontail/model/values/Complex32VectorValue.kt:141]: Perfect! I'd probably use a _IllegalArgumentException_ or a dedicated sub-class thereof. I consider calling these methods with mismatching dimensions a programmers error.

[the remark this answers is not in the document]
Comment (gabuzi) [reply src/main/kotlin/org/vitrivr/cottontail/model/values/Complex32VectorValue.kt]: done. Changed to IllegalArgumentException

**Pull request merged**